### PR TITLE
Issue #11514 - Cleanup `jetty.webapp.addServerClasses` property behavior for ee10/ee9/ee8

### LINKED ISSUE #11514: Start properties `jetty.webapp.addServerClasses` and `jetty.webapp.addSystemClasses` are not applied during ee8/ee9 deployments
**Jetty version(s)**
12.0.7

**Jetty Environment**
ee8 / ee9

**Java version/vendor** `(use: java -version)`
Any

**OS type/version**
Any

**Description**
If you setup an jetty.base with a `jetty.webapp.addServerClasses` (or `jetty.webapp.addSystemClasses`), then those properties do not get reflected in the ee9 / ee8 environments (it does work for the ee10 environments)

**How to reproduce?**

Setup a fresh jetty.base with the `ee9-demos` ...

```
$ mkdir demos-12-ee9
$ cd demos-12-ee9
$ java -jar ../../jetty-home-12.0.7/start.jar --add-module=http,ee9-deploy,ee9-demos
$ java -jar ../../jetty-home-12.0.7/start.jar --list-config | grep addServer
 jetty.webapp.addServerClasses = org.eclipse.jetty.logging.,${jetty.home.uri}/lib/logging/,org.slf4j.,${jetty.base.uri}/lib/bouncycastle/
```

Notice that we have a `addServerClasses` with the `org.eclipse.jetty.logging.` entry.

Now run with the server dump and look for the Serverclasses entries

```
$ java -jar ../../jetty-home-12.0.7/start.jar jetty.server.dumpAfterStart=true
|        |  +> Serverclasses EE9 Demo JSP WebApp@2ba5aa7a size=20
|        |  |  +> -org.eclipse.jetty.ee9.apache.
|        |  |  +> -org.eclipse.jetty.ee9.jsp.
|        |  |  +> -org.eclipse.jetty.ee9.servlet.DefaultServlet
|        |  |  +> -org.eclipse.jetty.ee9.servlet.NoJspServlet
|        |  |  +> -org.eclipse.jetty.ee9.servlet.StatisticsServlet
|        |  |  +> -org.eclipse.jetty.ee9.servlet.listener.
|        |  |  +> -org.eclipse.jetty.ee9.servlets.
|        |  |  +> -org.eclipse.jetty.ee9.websocket.api.
|        |  |  +> -org.eclipse.jetty.ee9.websocket.jakarta.client.JakartaWebSocketClientContainerProvider
|        |  |  +> -org.eclipse.jetty.ee9.websocket.jakarta.client.JakartaWebSocketShutdownContainer
|        |  |  +> -org.eclipse.jetty.ee9.websocket.jakarta.server.config.
|        |  |  +> -org.eclipse.jetty.ee9.websocket.server.
|        |  |  +> -org.eclipse.jetty.ee9.websocket.servlet.
|        |  |  +> -org.eclipse.jetty.jndi.
|        |  |  +> -org.eclipse.jetty.security.jaas.
|        |  |  +> org.eclipse.jdt.
|        |  |  +> org.eclipse.jetty.
|        |  |  +> org.eclipse.jetty.server.config.
|        |  |  +> org.eclipse.jetty.server.internal.
|        |  |  +> org.objectweb.asm.
|        |  +> Configurations EE9 Demo JSP WebApp@2ba5aa7a size=15
```

If we do the same thing, but with ee10 ...

```
$ mkdir demos-12-ee10
$ cd demos-12-ee10
$ java -jar ../../jetty-home-12.0.7/start.jar --add-module=http,ee10-deploy,ee10-demos
$ java -jar ../../jetty-home-12.0.7/start.jar --list-config | grep addServer
 jetty.webapp.addServerClasses = org.eclipse.jetty.logging.,${jetty.home.uri}/lib/logging/,org.slf4j.,${jetty.base.uri}/lib/bouncycastle/
```

And run with server dump ...

```
$ java -jar ../../jetty-home-12.0.7/start.jar jetty.server.dumpAfterStart=true
|        +> Serverclasses EE10 Demo Spec WebApp@32057e6 size=27
|        |  +> -org.eclipse.jetty.ee10.apache.
|        |  +> -org.eclipse.jetty.ee10.jaas.
|        |  +> -org.eclipse.jetty.ee10.jsp.
|        |  +> -org.eclipse.jetty.ee10.servlet.DefaultServlet
|        |  +> -org.eclipse.jetty.ee10.servlet.NoJspServlet
|        |  +> -org.eclipse.jetty.ee10.servlet.StatisticsServlet
|        |  +> -org.eclipse.jetty.ee10.servlet.listener.
|        |  +> -org.eclipse.jetty.ee10.servlets.
|        |  +> -org.eclipse.jetty.ee10.websocket.jakarta.client.JakartaWebSocketClientContainerProvider
|        |  +> -org.eclipse.jetty.ee10.websocket.jakarta.client.JakartaWebSocketShutdownContainer
|        |  +> -org.eclipse.jetty.ee10.websocket.jakarta.server.config.
|        |  +> -org.eclipse.jetty.ee10.websocket.server.
|        |  +> -org.eclipse.jetty.ee10.websocket.servlet.
|        |  +> -org.eclipse.jetty.jndi.
|        |  +> -org.eclipse.jetty.security.jaas.
|        |  +> -org.eclipse.jetty.websocket.api.
|        |  +> -org.eclipse.jetty.websocket.server.
|        |  +> file:///home/joakim/code/jetty/distros/bases/demos-12-ee10/lib/bouncycastle/
|        |  +> file:///home/joakim/code/jetty/distros/jetty-home-12.0.7/lib/logging/
|        |  +> org.eclipse.jdt.
|        |  +> org.eclipse.jetty.
|        |  +> org.eclipse.jetty.ee10.websocket.server.config.
|        |  +> org.eclipse.jetty.ee10.websocket.server.internal.
|        |  +> org.eclipse.jetty.logging.
|        |  +> org.eclipse.jetty.websocket.server.internal.
|        |  +> org.objectweb.asm.
|        |  +> org.slf4j.
|        +> Configurations EE10 Demo Spec WebApp@32057e6 size=15
```

You can see that the ee10 environment is working as expected with the `jetty.webapp.addServerClasses`

### PR DESCRIPTION
**Pull request by joakime**
Cleanup of how `jetty.webapp.addServerClasses` and `jetty.webapp.addSystemClasses` properties work in Jetty 12.

* Standardize all storage of configurations in the associated Environment attributes (we had some in Environment, some in Server)
* Deprecate static methods that take the Attributes or Server classes (as those were out of sync with what WebAppContext.setServer(Server) behavior was working.
* Add testing of new `WebAppContext.addServerClasses(String)`

This issue was discovered during the cross context dispatch testing, and is a critical piece of that testing.

Fixes #11514

### CODE REVIEW

File: jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java [hunk 1]
```diff
@@ -130,7 +130,12 @@
 public class ServletContextHandler extends ContextHandler
 {
     private static final Logger LOG = LoggerFactory.getLogger(ServletContextHandler.class);
-    public static final Environment __environment = Environment.ensure("ee10");
+    public static final Environment ENVIRONMENT = Environment.ensure("ee10");
+    /**
+     * @deprecated Use {@link ServletContextHandler#ENVIRONMENT} instead. will be removed in Jetty 12.1.0.
+     */
+    @Deprecated(since = "12.0.8", forRemoval = true)
+    public static final Environment __environment = ENVIRONMENT;
```
Comment (joakime) [inline jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:138]: This is just standardizing the public constant to the same way we have it in ee9/ee8.

File: jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java [hunk 2]
```diff
@@ -130,7 +130,12 @@
 public class ServletContextHandler extends ContextHandler
 {
     private static final Logger LOG = LoggerFactory.getLogger(ServletContextHandler.class);
-    public static final Environment __environment = Environment.ensure("ee10");
+    public static final Environment ENVIRONMENT = Environment.ensure("ee10");
+    /**
+     * @deprecated Use {@link ServletContextHandler#ENVIRONMENT} instead. will be removed in Jetty 12.1.0.
```
Comment (sbordet) [inline jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:135]: Remove the part related to when it will be removed.

File: jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java [hunk 3]
```diff
@@ -1383,24 +1383,62 @@ public MetaData getMetaData()
         return _metadata;
     }
 
-    public static void addServerClasses(Server server, String... pattern)
+    /**
+     * @param server ignored.
+     * @param patterns the patterns to add
+     * @deprecated use {@link #addServerClasses(String...)} instead, will be removed in Jetty 12.1.0
```
Comment (sbordet) [inline jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java:1389]: Remove when it will be removed.

File: jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java [hunk 4]
```diff
@@ -1383,24 +1383,62 @@ public MetaData getMetaData()
...
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("Adding {} ServerClasses: {}", ServletContextHandler.ENVIRONMENT, String.join(", ", patterns));
+        addClasses(__dftServerClasses, SERVER_SRV_CLASSES, patterns);
+    }
+
+    /**
+     * @param server ignored.
+     * @param patterns the patterns to add
+     * @deprecated use {@link #addSystemClasses(String...)} instead, will be removed in Jetty 12.1.0
```
Comment (sbordet) [inline jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java:1413]: Remove when it will be removed.

File: jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppContext.java [hunk 5]
```diff
@@ -1472,24 +1472,63 @@ public MetaData getMetaData()
...
+        if (LOG.isDebugEnabled())
+            LOG.debug("Adding {} ServerClasses: {}", ServletContextHandler.ENVIRONMENT, String.join(", ", patterns));
+        addClasses(__dftServerClasses, SERVER_SRV_CLASSES, patterns);
     }
 
-    public static void addSystemClasses(Attributes attributes, String... pattern)
+    /**
+     * @param attributes ignored.
+     * @param patterns the patterns to add
+     * @deprecated use {@link #addServerClasses(String...)} instead. will be removed in Jetty 12.1.0
```
Comment (sbordet) [inline jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppContext.java:1491]: Remove when it will be removed.

File: jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppContext.java [hunk 6]
```diff
@@ -1472,24 +1472,63 @@ public MetaData getMetaData()
...
+        if (LOG.isDebugEnabled())
+            LOG.debug("Adding {} SystemClasses: {}", ServletContextHandler.ENVIRONMENT, String.join(", ", patterns));
+
+        addClasses(__dftSystemClasses, SERVER_SYS_CLASSES, patterns);
+    }
+
+    /**
+     * @param attributes ignored.
+     * @param patterns the patterns to add
+     * @deprecated use {@link #addSystemClasses(String...)} instead. will be removed in Jetty 12.1.0
```
Comment (sbordet) [inline jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppContext.java:1516]: Remove when it will be removed.

**Review by sbordet — CHANGES_REQUESTED** [review]
I really don't like specifying future, possibly will-never-exist, Jetty versions in the code.
Other than that, LGTM.

File: jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java (hunk 2 above)
Comment (joakime) [reply jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:135]: I disagree, this is super common in use on open source java projects (spring-boot, spring-framework, dropwizard, tomcat, etc) and benefits developers.
Our own codebase already has many such usages before this PR.

Comment (sbordet) [reply jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:135]: @joakime then we should remove them all.

Spring has a policy (2 minor versions).

We do not know if we ever will have a 12.1.0, maybe we jump to 13 immediately due to the move to Java 21 required by JEE 11.

Or maybe we will have a policy similar to Spring.

Or maybe we retain it for sponsored reasons.

So the comment about 12.1.0 is wrong, in all cases, and must be removed.

Comment (gregw) [reply jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:135]: I'm on the fence here.   I like seeing the deprecated for removal.   I also don't mind being given info that will let me know I can keep using it (e.g. knowing it will always be there in 12.0.x).       But I also can see that naming a specific future version is not always correct (although we have frequently done it before and survived!).

So perhaps just say it is deprecated for removal "after 12.0.x", and that could be 12.1.0 or 13.0.0

File: jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java (hunk 3 above)
Comment (gregw) [inline jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java:1389]: The method that doesn't take a server does a different job.    That sets the default for all servers rather than this method, which sets the default for a specific server.

What is the replacement for setting the default for a specific server?  That should be linked here.

**Review by gregw — CHANGES_REQUESTED** [review]

File: jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java (hunk 2 above)
Comment (joakime) [reply jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:135]: > So the comment about 12.1.0 is wrong, in all cases, and must be removed.

I disagree with the "must be removed" part of your reply.
but I'm totally on board with having a minimum 2 minor releases policy.
That would change these references to say "12.2.0" instead, which means even if 13.0.0 gets released, they are removed, as 13.0.0 is after 12.2.0.

File: jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java (hunk 3 above)
Comment (joakime) [reply jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java:1389]: That's at the heart of what caused the issue.

The `Attributes` reference passed to `addServerClasses` was not used by the `WebAppContext.setServer(Server server)` call that needs the configuration.

Now with this PR the `add(Server|System)Classes` and `setServer(Server)` both use the environment specific `Attributes` locations.
Which works great, and even allows for wacky setups where 1 environment has a different setup than a different environment.

Do we have something like a jetty-core level `WebAppContext` (or something) that can be used as the `Server` level one?
I'm on the fence with the idea of having a `Server.add(Server|System)Classes` method which only has a `List<String>` (not the `ClassMatcher` option), but it feels like the wrong place for it.
Interestingly, while working this, I found a bug in the ee8 layer's XML where it attempted to use the `Server` instance to call `WebAppContext.addServerClasses(Server, String...)`, but due to XML order that `Server` didn't exist yet, which was simply worked around by using the Environment `Attributes` instead.

How about in a followup PR I move the 2 `ClassMatcher` impls (from ee10 & ee9) to jetty-util, and then put the `Attributes` neutral version in the new jetty-util code?

File: jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java (hunk 2 above)
Comment (sbordet) [reply jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:135]: @joakime then the policy is outlined elsewhere, e.g. in the docs.
Leave the code clean, without comments that are going to rot.

File: jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java (hunk 3 above)
Comment (sbordet) [reply jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java:1389]: @joakime note that the server classes are set on a static constant `ServletContextHandler.ENVIRONMENT`, so if you have 2 `Server`s in the same JVM, and you want different serverClasses configuration, you cannot.

File: jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppContext.java [hunk 7]
```diff
@@ -1504,6 +1543,6 @@ private static void addClasses(ClassMatcher matcher, String attribute, Attribute
         int l = classes.length;
         classes = Arrays.copyOf(classes, l + pattern.length);
         System.arraycopy(pattern, 0, classes, l, pattern.length);
-        attributes.setAttribute(attribute, classes);
+        ServletContextHandler.ENVIRONMENT.setAttribute(attributeKey, classes);
```
Comment (sbordet) [inline jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppContext.java:1546]: Why cannot system and server classes be a non-static field in `WebAppContext`?

The attributes are only used internally by `WebAppContext`, so there is no need to store them into a `static` constant, which will cause troubles in case of multiple `Server` instances in the same JVM.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java (hunk 2 above)
Comment (joakime) [reply jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:135]: The policy on deprecated doesn't exist in any documentation in Jetty.

Comment (sbordet) [reply jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:135]: Sure, let's add it, then. But it is orthogonal to this issue.

File: jetty-ee10/jetty-ee10-webapp/src/main/config/etc/jetty-ee10-webapp.xml [hunk 8]
```diff
@@ -3,7 +3,6 @@
 
 <Configure id="Server" class="org.eclipse.jetty.server.Server">
   <Call class="org.eclipse.jetty.ee10.webapp.WebAppContext" name="addSystemClasses">
-    <Arg><Ref refid="Server"/></Arg>
```
Comment (gregw) [inline jetty-ee10/jetty-ee10-webapp/src/main/config/etc/jetty-ee10-webapp.xml:6]: Why is this deleted?  This changes the configuration for all jetty servers in the JVM, rather than just the instance that this XML is configuring.   I agree it is unlikely that we have multiple servers when using jetty-home, but it is not impossible.

File: jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java [hunk 9]
```diff
@@ -1415,6 +1453,6 @@ private static void addClasses(ClassMatcher matcher, String attribute, Server se
         int l = classes.length;
         classes = Arrays.copyOf(classes, l + pattern.length);
         System.arraycopy(pattern, 0, classes, l, pattern.length);
-        server.setAttribute(attribute, classes);
+        ServletContextHandler.ENVIRONMENT.setAttribute(attributeKey, classes);
```
Comment (gregw) [inline jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java:1456]: Why are we setting this on an ENVIRONMENT?    Currently we should have system/server classes that can be set for:
 + all servers
 + a specific server
 + specific context

Perhaps this should be:
 + all servers
 + a specific server
 + a specific Environment
 + specific context

But this change looks to be just
 + all servers
 + a specific Environment
 + specific context

**Review by gregw — CHANGES_REQUESTED** [review]
I agree there is a problem and also that perhaps there is something more that we should do about configuring for an ENVIRONMENT, but I don't think we should give up configuring for a single server.

Other than the clean up, what was the actual bug that stopped EE8/EE9 seeing the server classes? Edit: sorry just saw your comment... reading your answer now....

**Comment by gregw** [comment]
@joakime thanks for [your comment](https://github.com/jetty/jetty.project/pull/11516#discussion_r1533887285) (which I initially missed).   It has explained a lot.

You have definitely found a problem and are on the way to a solution, but there are a couple of concepts that are not correct:

Currently (and with your PR), we have in each EE `WebAppContext`:
```java
    // System classes are classes that cannot be replaced by
    // the web application, and they are *always* loaded via
    // system classloader.
    public static final ClassMatcher __dftSystemClasses = new ClassMatcher(
        "java.",                            // Java SE classes (per servlet spec v2.5 / SRV.9.7.2)
        "javax.",                           // Java SE classes (per servlet spec v2.5 / SRV.9.7.2)
        "jakarta.",                         // Jakarta classes (per servlet spec v5.0 / Section 15.2.1)
        "org.xml.",                         // javax.xml
        "org.w3c."                          // javax.xml
    );

    // Server classes are classes that are hidden from being
    // loaded by the web application using system classloader,
    // so if web application needs to load any of such classes,
    // it has to include them in its distribution.
    public static final ClassMatcher __dftServerClasses = new ClassMatcher(
        "org.eclipse.jetty."                // hide jetty classes
    );
```

But because these statics are on EE classes, they are not server wide.  They are effectively only in the `Environment` that loads the class, so they are Env defaults not server defaults.
Switching to use the `Environment` as the `Attributes` is not really the right thing to do, as these attributes have exactly the same scope as the statics on the `WebAppContext` classes.

There is also the issue that EE10 `WebAppContext` uses:
```java
    public static final String SERVER_SYS_CLASSES = "org.eclipse.jetty.webapp.systemClasses";
    public static final String SERVER_SRV_CLASSES = "org.eclipse.jetty.webapp.serverClasses";
```

Whilst  EE9 `WebAppContext` uses names scoped to the environment:
```java
    public static final String SERVER_SYS_CLASSES = "org.eclipse.jetty.ee9.webapp.systemClasses";
    public static final String SERVER_SRV_CLASSES = "org.eclipse.jetty.ee9.webapp.serverClasses";
```
Having an attribute name scope to an environment inside an environment scoped Attributes is at least duplicitous if not an anti-pattern.

So you have definitely identified a bit of a mess!    But I think we need bigger steps to fix it than what you have here.    However, I think you are on the right course when you say:

> move the 2 ClassMatcher impls (from ee10 & ee9) to jetty-util, and then put the Attributes neutral version in the new jetty-util code

I think we should do that now (or perhaps to a `jetty-ee-common` module )and build up from there so we can have system/server classes specified for:
 + all servers in the JVM (stored as ClassMatcher static field on some `jetty-ee-common` class)
 + a specific server (stored as an attribute on the Server instance)
 + a specific environment (stored either as attribute on the `Environment` or as a ClassMatcher static field on the EE specific WebAppContext class 
 + a specific context (stored as a ClassMatcher  field on the EE specific WebAppContext class 

We will need the `jetty-eeX-webapp` modules to depend on a `jetty-ee-webapp` module

This is a bit of work, but we have a few days to get it done for this release cycle.... if we don't finish, we can bump to the next.

Other than that, the cleanups of naming etc. in this PR are good.

File: jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java (hunk 3 above)
Comment (joakime) [reply jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppContext.java:1389]: The ServletContextHandler.ENVIRONMENT (that exists in ee10, ee9, ee8) are JVM statics ATM (in HEAD).
I think changing that is out of scope for this PR, but I can see that possibly being a useful concept in the long run.

File: jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppContext.java (hunk 7 above)
Comment (joakime) [reply jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebAppContext.java:1546]: This method is used for configuring all WebAppContext's that get created for this environment in a common way.
There is already WebAppContext instance specific ways of configuring the ClassMatcher used by the WebApp.
Take a look at `WebAppContext.setServer(Server)` for where the instance specific ClassMatcher is configured.

**Comment by gregw** [comment]
@joakime I've started a branch where I'm moving a few files from eeX to a new jetty-core/jetty-ee module.  Give me a day or two to progress that, then you can perhaps merge with your branch and use it...

**Comment by gregw** [comment]
@joakime my first pass at this is here: org.eclipse.jetty.ee.WebappProtectedClasses     probably a bit more work needed.

Edit: I mean fix/12.0.x/addserverclasses-ee9-gw

**Comment by gregw** [comment]
See #11566 with ClassMatcher in util and a common ee module

**Comment by joakime** [comment]
Closing as PR #11566 completed this need
